examples/source_separation/conv_tasnet/metrics.py

```diff
@@ -0,0 +1,116 @@
+
+    reference_pow = reference.pow(2).sum(axis=2, keepdim=True)
+    mix_pow = (estimate * reference).sum(axis=2, keepdim=True)
+    scale = mix_pow / (reference_pow + epsilon)
+
+    reference = scale * reference
+    error = estimate - reference
+
+    reference_pow = reference.pow(2).sum(axis=2)
+    error_pow = error.pow(2).sum(axis=2)
+
+    return 10 * torch.log10(reference_pow) - 10 * torch.log10(error_pow)
+
+
+def neg_si_snr(
+    estimate: torch.Tensor, reference: torch.Tensor, epsilon=1e-8
+) -> torch.Tensor:
+    """Computes **NEGATIVE** scale-invariant source-to-noise ratio.
+
+    The smaller the returned value is, the better the signal reconstruction is.
+
+    Args:
+        estimate (torch.Tensor): Estimtaed value. Shape: [batch, speaker, time frame]
+        reference (torch.Tensor): Reference value. Shape: [batch, speaker, time frame]
+        epsilon (float): constant value used to stabilize division.
+
+    Returns:
+        torch.Tensor: **NEGATIVE** scale-invariant source-to-distortion ratio.
+        Shape: [batch, speaker]
+    """
+    return -si_snr(estimate, reference, epsilon)
+
+
+class PITLoss(torch.nn.Module):
+    """Utterance-level permutation invariant Loss
+
+    Computes the minimum possible loss value over the permutations of the speaker.
+
+    Args:
+        loss_func (function):
+            Loss function with signature of
+            (input: torch.Tensor, target: torch.Tensor) -> torch.Tensor
+            where input Tensors are shape of (batch, #of speakers, frame) and
+            the output Tensor retains batch dimention (batch, ...).
+
+    References:
+        Multi-talker Speech Separation with Utterance-level Permutation Invariant Training of
+        Deep Recurrent Neural Networks
+
+        https://arxiv.org/abs/1703.06284
+    """
+
+    def __init__(self, loss_func=neg_si_snr):
+        super().__init__()
+        self.loss_func = loss_func
+
+    def forward(
+        self, estimate: torch.Tensor, target: torch.Tensor, reduction: str = "mean"
+    ) -> torch.Tensor:
+        """Compute utterance-level PIT Loss
+
+        Args:
+            estimate (torch.Tensor): Estimated signal reconstruction.
+                Shape: [bacth, speaker, time frame]
+            target (torch.Tensor): Target signal.
+                Shape: [bacth, speaker, frame]
+            reduction (str): Reduction method to aggregate the criterion values over batch dimension and
+                other dimensions that criterion returns.
+
+        Returns:
+            torch.Tensor: Criterion based on the speaker permutation that give the minimum criterion.
+        """
+        assert estimate.shape == target.shape
+        assert reduction in ["mean", "sum"]
+
+        num_speakers = target.shape[1]
+
+        min_metric = None
+        min_metric_value = float("inf")
+        for idx in permutations(range(num_speakers)):
+            metric = self.loss_func(estimate, target[:, idx, :])
```

Comment:
This applies the shuffling to all the samples in batch. Each samples should be shuffled separately.